airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt

```diff
@@ -0,0 +1,129 @@
+package io.airbyte.cdk.test.util
+
+import com.fasterxml.jackson.databind.JsonNode
+import io.airbyte.cdk.Operation
+import io.airbyte.cdk.command.CliRunner
+import io.airbyte.cdk.command.ConfigurationJsonObjectBase
+import io.airbyte.cdk.output.BufferingOutputConsumer
+import io.airbyte.protocol.models.Jsons
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
+import io.micronaut.context.RuntimeBeanDefinition
+import java.io.InputStream
+import java.io.PipedInputStream
+import java.io.PipedOutputStream
+import java.io.PrintStream
+import java.io.PrintWriter
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.Future
+import javax.inject.Singleton
+
+// This whole file is very wishy-washy, until we figure out the exact
+// micronaut stuff. But it's directionally correct, and I'd rather nail down
+// the actual test stuff first.
+interface DestinationProcess {
+    fun sendMessage(message: AirbyteMessage)
+
+    /**
+     * Return all messages from the destination. When the destination exits, the last message
+     * MUST be `null`.
+     *
+     * (mediocre interface, just for demo purposes)
+     */
+    fun readMessages(): List<AirbyteMessage?>
+
+    fun waitUntilDone()
+}
+
+abstract class DestinationProcessFactory {
+    abstract fun createDestinationProcess(
+        command: String,
+        catalog: ConfiguredAirbyteCatalog? = null,
+    ): DestinationProcess
+}
+
+class NonDockerizedDestination(
+    command: String,
+    config: ConfigurationJsonObjectBase?,
+    catalog: ConfiguredAirbyteCatalog?,
+): DestinationProcess {
+    private val destinationStdin = PipedInputStream()
+    private val destinationStdinPipe = PrintWriter(PipedOutputStream(destinationStdin))
+    // TODO CliRunner.runDestination() is blocking, so we can't use it directly
+    //   but this is objectively wrong - we need to be able to stream the output
+    //   throughout the sync, whereas right now we have to do
+    //   destination.get().messages(). Which blocks unti the whole destination
+    //   finishes running.
+    private val destination: Future<BufferingOutputConsumer> =
+        CompletableFuture.supplyAsync {
+            CliRunner.runDestination(
+                command,
+                config = config,
+                catalog = catalog,
+                // TODO is this really the right way to achieve this?
```

Comment:
@johnny-schmidt before I go _too_ far down this path - this (+the related plumbing code in CliRunner + AirbyteConnectorRunner) seems like a nicer way to inject custom beans to the destination? I.e. this way, test authors don't need to define a new InputStreamFactory + do property dancing

but lmk if you have a different idea for doing this

Reply:
If the idea is that we pass in config & catalog explicitly, would it make more sense to provide the input data explicitly also? Just wrap it in an serializing InputStream?

Reply:
maaaybe, though I'm leaning towards no. There are tests that want to interleave messages and test logic, a la
```
sendSomeRecords()
verifySomething()
sendMoreRecords()
endSync()
verifySomethingElse()
```

At which point I'd rather wrap it in the destination process via the `sendMessage` function, rather than exposing the pipes to callers? I'm imagining something along these lines, which doesn't feel great?
```
val pipe = PipedOutputStream()
val pipeInputStream = PipedInputStream(pipe)
val dest = DestinationProcess(pipeInputStream)
messages1.each { pipe.write(serialize(it)) }
verifySomething()
messages2.each { pipe.write(serialize(it)) }
// flush all messages
pipe.close()
// this will magically cause the destination to shutdown
pipeInputStream.close()
```

plus I think we get most of the benefit by providing a utility method wrapping the destination process?
```
fun runSync(factory, config, catalog, messages) {
  val dest = factory.start(config, catalog)
  messages.each { dest.sendMessage(it) }
  return dest.readMessages()
}
```